### PR TITLE
New rule S6566: Always use DateTimeOffset instead of DateTime

### LINKED ISSUE #7085: New rule S6566: Always use `DateTimeOffset` instead of `DateTime`
Implement new rule [S6566: Use DateTimeOffset instead of DateTime](https://github.com/SonarSource/rspec/pull/1717).

This rule should raise on the `DateTime` class is used.

- [ ] Implement Rule S6566 for C#
- [ ] Implement Rule S6566 for VB.NET

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #7085

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.cs
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.cs:61]: I am not 100% convinced we should raise on the methods.
Why?

- For the parse methods if the string contains only the information for a date, you cannot create a valid DateTimeOffset.
- For the specifykind the reasoning is similar, if you have a date, at this point when you are specifying the kind you might not have the offset information.

In my opinion if we only raise for constructors and the well known fields like, Now, UTCNow, Today, etc. We will ensure that the sources of DateTime objects are converted, and the change will propagate trough the whole application and everything will be changed to DateTimeOffset.

Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.cs:61]: Can you provide an example of a DateTime string that would be not parsed correctly by the DateTimeOffset.Parse() method?
I tried playing a bit with[ this example](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBfsQAwAEGECApgMoLKwDmdAvHQCIaAVgD0NYqMI0aADgEBuCgDcIyBk2YAVKGGZ86AEU069AOgAKagM7MAFIxbtOMLgEol5VesfbdzAHkAMyDbBANjFlNAkLDLG3tfZ24PCipiAE4HE39U8mos32jg0OYEDyA===) and they appear to be working similarly.

Comment (csaba-sagi-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.cs:61]: That is exactly the problem there. Is it a conscious decision? I am afraid people will just replace DateTime with DateTimeOffset and will have always 0 as the offset. Whan in my opinion the rule should achieve is to "open the eyes" of the devs and tell them to have a conscious decision and use the offset to define where the time points to.
If the offset is always 0 because the string does not contain a valid offset, how is that DateTimeOffset objects different from a simple DateTime with Local kind?

Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.cs:61]: For DateTime.Parse the kind is Unspecified, so the information is ambiguous from the very beginning. IMO when no offset is specified it is safe to assume that we are talking about UTC, unfortunately, Microsoft does not agree:

> However, for [DateTime](https://learn.microsoft.com/en-us/dotnet/api/system.datetime) values whose [Kind](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.kind) property is [DateTimeKind.Unspecified](https://learn.microsoft.com/en-us/dotnet/api/system.datetimekind#system-datetimekind-unspecified), these two conversion methods produce a [DateTimeOffset](https://learn.microsoft.com/en-us/dotnet/api/system.datetimeoffset) value whose offset is that of the local time zone.

[Source](https://learn.microsoft.com/en-us/dotnet/standard/datetime/converting-between-datetime-and-offset#conversions-from-datetime-to-datetimeoffset)

So I think I agree with you, I'll remove the parsing methods from here.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.cs
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.cs:24]: The message is about using one type instead of another. So I'd only highlight the type name / declaration, rather than the whole member access expression (it's usage).

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs:6]: These constructors are available in the .NET Framework and they don't use any kind of expression that isn't available in earlier C# versions. Why is it placed in this test file?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs:11]: What do these 3 lines test?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.cs
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.cs:1]: Add a test case for the `nameof` operator:
`var typeName = nameof(DateTime);`

File: analyzers/src/SonarAnalyzer.Common/Rules/UseDateTimeInsteadOfDateTimeOffsetBase.cs
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseDateTimeInsteadOfDateTimeOffsetBase.cs:67]: The two registration can be merged if you
1. Register for `Language.SyntaxKind.IdentifierName` instead.
2. Check if the name is `DateTime` or `System.DateTime` (or `Date` in VB).
3. If it is then do a symbolic check to see if it's actually `System.DateTime`.
4. Mark certain edge cases (e.g. `System.DateTime` is aliased with the using directive) as FN.

This will help in covering some of the additional scenarios: method parameters, field and property declarations.

**Review by zsolt-kolbay-sonarsource — CHANGES_REQUESTED** [review]
Review Round 1

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/UseDateTimeInsteadOfDateTimeOffset.cs
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Rules/UseDateTimeInsteadOfDateTimeOffset.cs:28]: Why? Couldn't you check for name Date and DateTime?
And couldn't you move the check itself to the base class and have an array of names to check for, that would be language specific?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.vb
Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.vb:49]: It was helpful for me to have the whole list and decide whether we should raise or not on a case-by-case basis. IMO we can keep the list there; it's not that long and will give an overall view of what's available without going back and forth with the docs

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs
Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs:6]: If run those 3 constructors in .NET Framework they results in:
`build error [CS1503]: Argument 8: cannot convert from 'int' to 'System.Globalization.Calendar'`
or
`build error [CS1729]: 'DateTime' does not contain a constructor that takes 10 arguments`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs
Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs:11]: Constructor not recognized in .NET Framework, [UnixEpoch](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.unixepoch) introduced in Core 2.1 or Standard 2.1 and the last span was initialized to test the [TryFormat](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.tryformat) method

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/UseDateTimeInsteadOfDateTimeOffset.cs
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.VisualBasic/Rules/UseDateTimeInsteadOfDateTimeOffset.cs:28]: Yes, we can (now). Before our discussion, we had also noncompliant non-static properties/fields, so checking the name was not feasible. I'll update the logic now 👍

File: analyzers/src/SonarAnalyzer.Common/Rules/UseDateTimeInsteadOfDateTimeOffsetBase.cs
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseDateTimeInsteadOfDateTimeOffsetBase.cs:67]: This way you will miss all the implicit object creations if I'm not mistaken:
`DateTime date = new(1);`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs:11]: `UnixEpoch` is fine, but the 3 lines following that will not trigger the rule. If you want to test whether the rule will raise an issue on instance method calls, then that can be done in the common (.NET Framework) test cases.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseDateTimeInsteadOfDateTimeOffsetBase.cs
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseDateTimeInsteadOfDateTimeOffsetBase.cs:71]: Yes you are right indeed. Still I'm not sure because we don't want to raise on every System.DateTime identifier:
```cs
DateTime.IsLeapYear(1);
        DateTime.Parse("06/01/1993");
        DateTime.ParseExact("06/01/1993", "dd/MM/yyyy", null);
        DateTime.SpecifyKind(date, DateTimeKind.Local);
        var typeName = nameof(DateTime); // Compliant
```

So you need to do the same checks in reverse to see whether they are indeed the noncompliant creation/access expression. IDK, I think it's fine to register directly for both and have a simple logic that checks the children instead of the parents/context.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs
Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseDateTimeInsteadOfDateTimeOffset.CSharp9.cs:11]: Wait I thought with "3 lines" you were referring to 8 10 and 11 (the one highlighted in the comments above).
If you are referring to 11 12 and 13(?) then 11 is just to create the span needed for the TryFormat method line 12 (we are only making sure that this is indeed compliant). Line 13 is just AddMicroseconds method, not supported in .NET Framework

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseDateTimeInsteadOfDateTimeOffset.cs
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UseDateTimeInsteadOfDateTimeOffset.cs:34]: You can use `Language.Syntax.NodeExpression` to get the expression in the base class and just call GetLocation on it. So this method will not be needed.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseDateTimeInsteadOfDateTimeOffset.cs
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UseDateTimeInsteadOfDateTimeOffset.cs:28]: You can have an abstract property in the base class with the possible date time names, and then you can do the check in the base class and will not need this method.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseDateTimeInsteadOfDateTimeOffset.cs
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UseDateTimeInsteadOfDateTimeOffset.cs:31]: Once S6562 gets merged you can use `Language.Syntax.ObjectCreationTypeIdentifier` to get the identifier and you will be able to do the type check in the base class.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseDateTimeInsteadOfDateTimeOffsetBase.cs
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseDateTimeInsteadOfDateTimeOffsetBase.cs:71]: There are a lot of Identifiers in a source code, so in my opinion, it is better to have the two separate registrations. It should perform better.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
I left a couple of comments there.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseDateTimeInsteadOfDateTimeOffset.cs
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UseDateTimeInsteadOfDateTimeOffset.cs:30]: The PR for S6562 is approved, once the pipeline is green, it will be merged. I suggest to rebase on it and get rid of this abstract method.

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM. Please rebase once S6562 is merged to get rid of the GetType method.

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**
